File: core/src/test/java/ru/pocketbyte/locolaser/LocoLaserTest.java
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser;

import org.json.simple.JSONObject;
import org.json.simple.parser.ParseException;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import ru.pocketbyte.locolaser.config.Config;
import ru.pocketbyte.locolaser.config.WritingConfig;
import ru.pocketbyte.locolaser.config.platform.PlatformConfig;
import ru.pocketbyte.locolaser.config.source.Source;
import ru.pocketbyte.locolaser.config.source.SourceConfig;
import ru.pocketbyte.locolaser.resource.PlatformResources;
import ru.pocketbyte.locolaser.resource.entity.*;
import ru.pocketbyte.locolaser.summary.FileSummary;
import ru.pocketbyte.locolaser.summary.Summary;
import ru.pocketbyte.locolaser.utils.JsonParseUtils;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.*;

import static org.junit.Assert.*;

/**
 * @author Denis Shurygin
 */
public class LocoLaserTest {

    @Rule
    public TemporaryFolder tempFolder= new TemporaryFolder();

    private Config config;
    private MockPlatformConfig platformConfig;
    private MockPlatformResources platformResources;

    private MockSource source;
    private MockSourceConfig sourceConfig;

    private ResMap mResMap;

    @Before
    public void init() throws IOException {
        File workDir = tempFolder.newFolder();
        System.setProperty("user.dir", workDir.getCanonicalPath());

        Summary.setFactory(new Summary.Factory() {
            @Override
            public Summary loadSummary(Config config) {
                return null;
            }
        });

        ResLocale localeEn = new ResLocale();
        localeEn.put(buildItem("key1", "valueEn1"));
        localeEn.put(buildItem("key2", "valueEn2"));

        ResLocale localeRu = new ResLocale();
        localeRu.put(buildItem("key1", "valueRu1"));
        localeRu.put(buildItem("key2", "valueRu2"));

        mResMap = new ResMap();
        mResMap.put("en", localeEn);
        mResMap.put("ru", localeRu);

        platformResources = new MockPlatformResources(null, null);
        platformConfig = new MockPlatformConfig("mockPlatform", platformResources);

        source = new MockSource(new ResMap(mResMap), System.currentTimeMillis());
        sourceConfig = new MockSourceConfig("mockSource", source, new HashSet<>(Arrays.asList("en", "ru")));

        config = new Config();
        config.setFile(new File(workDir, "config.json"));
        config.setPlatform(platformConfig);
        config.setSourceConfig(sourceConfig);

        // Write config file to make it not empty
        PrintWriter writer = new PrintWriter(config.getFile());
        writer.write("{}");
        writer.flush();
        writer.close();
    }

    @After
    public void deInit() {
        Summary.setFactory(null);
    }

    // ====================================================
    // Test invalid config

    @Test(expected = IllegalArgumentException.class)
    public void testNullConfig() {
        LocoLaser.localize(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testNullPlatform() {
        config.setPlatform(null);
        LocoLaser.localize(config);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testNullSources() {
        config.setSourceConfig(null);
        LocoLaser.localize(config);
    }

    // ====================================================
    // Test conflict strategy

    @Test
    public void testConflictStrategyRemoveLocal() {
        config.setConflictStrategy(Config.ConflictStrategy.REMOVE_PLATFORM);

        String locale = "en";
        String newKey = "newKey";
        // Check if original map doesn't contain new value
        assertFalse(mResMap.get(locale).containsKey(newKey));

        ResLocale resLocale = new ResLocale();
        resLocale.put(buildItem(newKey, "value"));

        platformResources.mMap = new ResMap();
        platformResources.mMap.put(locale, resLocale);

        assertTrue(LocoLaser.localize(config));
        assertEquals(platformResources.mMap, source.mMap);
    }

    @Test
    public void testConflictStrategyKeepNewLocal() {
        config.setConflictStrategy(Config.ConflictStrategy.KEEP_NEW_PLATFORM);

        String locale = "en";
        String newKey = "newKey";
        // Check if original map doesn't contain new value
        assertFalse(mResMap.get(locale).containsKey(newKey));

        ResLocale resLocale = new ResLocale();
        resLocale.put(buildItem(newKey, "value"));

        platformResources.mMap = new ResMap();
        platformResources.mMap.put(locale, resLocale);

        assertTrue(LocoLaser.localize(config));
        assertNotEquals(platformResources.mMap, source.mMap);

        platformResources.mMap.get(locale).remove(newKey);
        assertEquals(platformResources.mMap, source.mMap);
    }

    @Test
    public void testConflictStrategyExportNewLocal() {
        config.setConflictStrategy(Config.ConflictStrategy.EXPORT_NEW_PLATFORM);

        String locale = "en";
        String newKey = "newKey";
        // Check if original map doesn't contain new value
        assertFalse(mResMap.get(locale).containsKey(newKey));

        ResLocale resLocale = new ResLocale();
        resLocale.put(buildItem(newKey, "value"));

        platformResources.mMap = new ResMap();
        platformResources.mMap.put(locale, resLocale);

        assertTrue(LocoLaser.localize(config));
        assertEquals(platformResources.mMap, source.mMap);

        assertTrue(source.mMap.get(locale).containsKey(newKey));
    }


    @Test
    public void testConflictStrategyExportMissedValues() {
        config.setConflictStrategy(Config.ConflictStrategy.EXPORT_NEW_PLATFORM);

        String missedKey = "key1";
        String locale = "ru";
        Source.ValueLocation location = new Source.ValueLocation(source) {};

        platformResources.mMap = new ResMap(mResMap);
        assertNotNull(source.mMap.get(locale).remove(missedKey));

        source.mMissedValues = Collections.singletonList(
                new Source.MissedValue(
                        missedKey, locale,
                        Quantity.OTHER, location));

        assertTrue(LocoLaser.localize(config));
        assertEquals(platformResources.mMap, source.mMap);

        ResItem missedItem = source.mMap.get(locale).get(missedKey);
        assertNotNull(missedItem);
        assertTrue(missedItem.valueForQuantity(Quantity.OTHER).getLocation() == location);
    }

    // ====================================================
    // Other

    @Test
    public void testSimpleReadAndWrite() {
        assertNull(platformResources.mMap);
        assertNotNull(source.mMap);

        assertTrue(LocoLaser.localize(config));

        assertEquals(platformResources.mMap, source.mMap);
    }

    @Test
    public void testWritingConfigUsage() {
        assertNull(platformResources.mWritingConfig);
        assertNotNull(config.writingConfig);

        assertTrue(LocoLaser.localize(config));

        assertTrue(platformResources.mWritingConfig == config.writingConfig);
    }

    @Test
    public void testLocalizationNotNeeded() throws ParseException, IOException {
        prepareStateWhenLocalizationNotNeededBecauseNoChanges();

        assertTrue(LocoLaser.localize(config));
        assertNull(platformResources.mMap);
    }

    @Test
    public void testForceImport() throws ParseException, IOException {
        prepareStateWhenLocalizationNotNeededBecauseNoChanges();

        config.setForceImport(true);
        assertTrue(LocoLaser.localize(config));
        assertNotNull(platformResources.mMap);
        assertEquals(platformResources.mMap, source.mMap);
    }

    @Test
    public void testConfigChanged() throws ParseException, IOException {
        prepareStateWhenLocalizationNotNeededBecauseNoChanges();

        // Override config file
        PrintWriter writer = new PrintWriter(config.getFile());
        writer.write("{\"key\":\"wrong\"}");
        writer.flush();
        writer.close();

        assertTrue(LocoLaser.localize(config));
        assertNotNull(platformResources.mMap);
        assertEquals(platformResources.mMap, source.mMap);
    }

    @Test
    public void testSourceChanged() throws ParseException, IOException {
        prepareStateWhenLocalizationNotNeededBecauseNoChanges();

        source.mModifiedDate++;

        assertTrue(LocoLaser.localize(config));
        assertNotNull(platformResources.mMap);
        assertEquals(platformResources.mMap, source.mMap);
    }

    @Test
    public void testLocalResourceChanged() throws ParseException, IOException {
        prepareStateWhenLocalizationNotNeededBecauseNoChanges();

        String changedLocale = "en";
        FileSummary localeSummary = platformResources.mSummaryMap.get(changedLocale);

        platformResources.mSummaryMap.put(changedLocale, new FileSummary(localeSummary.bytes + 1, "hash_en"));

        assertTrue(LocoLaser.localize(config));
        assertNotNull(platformResources.mMap);
        assertNotEquals(platformResources.mMap, source.mMap);
        assertEquals(1, platformResources.mMap.size());
        assertEquals(platformResources.mMap.get(changedLocale), source.mMap.get(changedLocale));
    }

    @Test
    public void testSourceClose() {
        assertTrue(LocoLaser.localize(config));
        assertTrue(source.isClosed);
    }

    @Test
    public void testSaveSummary() throws ParseException, IOException {
        platformResources.mSummaryMap = new HashMap<>(2);
        platformResources.mSummaryMap.put("en", new FileSummary(2300000, "hash_en"));
        platformResources.mSummaryMap.put("ru", new FileSummary(2123123, "hash_ru"));

        String jsonString = "{" +
                "\"" + Summary.CONFIG_FILE + "\":" + new FileSummary(12312, "old_config").toJson() + "," +
                "\"" + Summary.SOURCE_MODIFIED_DATE + "\":"+ (source.getModifiedDate() + 1) +"," +
                "\"" + Summary.RESOURCE_FILES + "\":{" +
                "\"en\":" + new FileSummary(23213, "old_en").toJson() + "," +
                "\"ru\":" + new FileSummary(24563, "old_ru").toJson() + "}" +
                "}";

        File tempDir = config.getTempDir();
        if (!tempDir.exists())
            assertTrue(tempDir.mkdirs());
        File summaryFile = new File(tempDir, Summary.SUMMARY_FILE_NAME);

        PrintWriter writer = new PrintWriter(summaryFile);
        writer.write(jsonString);
        writer.flush();
        writer.close();

        Summary.setFactory(null);

        //Check if summary saved
        Summary summary = Summary.loadSummary(config);
        assertEquals(new FileSummary(12312, "old_config"), summary.getConfigSummary());
        assertEquals(new FileSummary(23213, "old_en"), summary.getResourceSummary("en"));
        assertEquals(new FileSummary(24563, "old_ru"), summary.getResourceSummary("ru"));
        assertEquals(source.getModifiedDate() + 1, summary.getSourceModifiedDate());

        assertTrue(LocoLaser.localize(config));

        //Check if summary changed
        summary = Summary.loadSummary(config);
        assertEquals(new FileSummary(config.getFile()), summary.getConfigSummary());
        assertEquals(new FileSummary(2300000, "hash_en"), summary.getResourceSummary("en"));
        assertEquals(new FileSummary(2123123, "hash_ru"), summary.getResourceSummary("ru"));
        assertEquals(source.getModifiedDate(), summary.getSourceModifiedDate());
    }

    // ====================================================
    // Private classes and methods

    private ResItem buildItem(String key, String value) {
        ResItem item = new ResItem(key);
        item.addValue(new ResValue(value, null));
        return item;
    }

    private Summary prepareStateWhenLocalizationNotNeededBecauseNoChanges() throws IOException, ParseException {
        platformResources.mSummaryMap = new HashMap<>(2);
        platformResources.mSummaryMap.put("en", new FileSummary(23, "hash_en"));
        platformResources.mSummaryMap.put("ru", new FileSummary(2123123, "hash_ru"));

        String jsonString = "{" +
                "\"" + Summary.CONFIG_FILE + "\":" + new FileSummary(config.getFile()).toJson() + "," +
                "\"" + Summary.SOURCE_MODIFIED_DATE + "\":"+ source.getModifiedDate() +"," +
                "\"" + Summary.RESOURCE_FILES + "\":{" +
                "\"en\":" + platformResources.mSummaryMap.get("en").toJson() + "," +
                "\"ru\":" + platformResources.mSummaryMap.get("ru").toJson() + "}" +
                "}";

        final Summary summary = new Summary(tempFolder.newFile(),
                (JSONObject) JsonParseUtils.JSON_PARSER.parse(jsonString));

        Summary.setFactory(new Summary.Factory() {
            @Override
            public Summary loadSummary(Config config) {
                return summary;
            }
        });

        return summary;
    }

    private class MockPlatformConfig implements PlatformConfig {

        private String mName;
        private PlatformResources mResources;

        MockPlatformConfig(String name, PlatformResources resources) {
            mName = name;
            mResources = resources;
        }

        @Override
        public String getType() {
            return mName;
        }

        @Override
        public PlatformResources getResources() {
            return mResources;
        }

        @Override
        public File getTempDir() {
            return new File(System.getProperty("user.dir"), "./temp/");
        }
    }

    private class MockPlatformResources implements PlatformResources {

        private ResMap mMap;
        private Map<String, FileSummary> mSummaryMap;
        private WritingConfig mWritingConfig;

        MockPlatformResources(ResMap resMap, Map<String, FileSummary> summaryMap) {
            mMap = resMap;
            mSummaryMap = summaryMap;
        }

        @Override
        public ResMap read(Set<String> locales) {
            ResMap resMap = new ResMap();
            if (mMap != null) {
                for (String locale : locales) {
                    resMap.put(locale, new ResLocale(mMap.get(locale)));
                }
            }
            return resMap;
        }

        @Override
        public void write(ResMap map, WritingConfig writingConfig) throws IOException {
            mMap = new ResMap(map);
            mWritingConfig = writingConfig;
        }

        @Override
        public FileSummary summaryForLocale(String locale) {
            if (mSummaryMap != null)
                return mSummaryMap.get(locale);
            return null;
        }
    }

    private static class MockSourceConfig implements SourceConfig {

        private String mType;
        private Source mSource;
        private Set<String> mLocales;

        MockSourceConfig(String type, Source source, Set<String> locales) {
            mType = type;
            mSource = source;
            mLocales = locales;
        }

        @Override
        public String getType() {
            return mType;
        }

        @Override
        public Source open() {
            return mSource;
        }

        @Override
        public Set<String> getLocales() {
            return mLocales;
        }
    }

    private class MockSource extends Source {

        private boolean isClosed = false;
        private ResMap mMap;
        private List<MissedValue> mMissedValues;
        private long mModifiedDate;

        public MockSource(ResMap resMap, long modifiedDate) {
            super(null);
            mMap = resMap;
            mModifiedDate = modifiedDate;
        }

        @Override
        public long getModifiedDate() {
            return mModifiedDate;
        }

        @Override
        public ReadResult read() {
            return new ReadResult(new ResMap(mMap), mMissedValues);
        }

        @Override
        public void write(ResMap resMap) {
            if (mMap == null)
                mMap = new ResMap(resMap);
            else
                mMap.merge(resMap);
        }

        @Override
        public void close() {
            if (isClosed)
                throw new IllegalStateException("Source should be closed only once");
            isClosed = true;
        }
    }
}

File: core/src/test/java/ru/pocketbyte/locolaser/LocoLaserTest.kt
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser

import org.junit.Assert.*
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import ru.pocketbyte.locolaser.config.Config
import ru.pocketbyte.locolaser.config.ExtraParams
import ru.pocketbyte.locolaser.config.resources.ResourcesConfig
import ru.pocketbyte.locolaser.config.trimUnsupportedQuantities
import ru.pocketbyte.locolaser.entity.Quantity
import ru.pocketbyte.locolaser.resource.Resources
import ru.pocketbyte.locolaser.resource.entity.*
import ru.pocketbyte.locolaser.resource.formatting.FormattingType
import ru.pocketbyte.locolaser.resource.formatting.NoFormattingType
import java.io.File
import java.io.IOException
import java.io.PrintWriter

/**
 * @author Denis Shurygin
 */
class LocoLaserTest {

    @Rule @JvmField
    var tempFolder = TemporaryFolder()

    private lateinit var config: Config
    private lateinit var platformConfig: MockResourcesConfig
    private lateinit var platformResources: MockResources

    private lateinit var source: MockSource
    private lateinit var sourceConfig: MockSourceConfig

    private lateinit var mResMap: ResMap

    @Before
    @Throws(IOException::class)
    fun init() {
        val workDir = tempFolder.newFolder()
        System.setProperty("user.dir", workDir.canonicalPath)

        val localeEn = ResLocale()
        localeEn.put(buildItem("key1", "valueEn1"))
        localeEn.put(buildItem("key2", "valueEn2"))

        val localeRu = ResLocale()
        localeRu.put(buildItem("key1", "valueRu1"))
        localeRu.put(buildItem("key2", "valueRu2"))

        mResMap = ResMap()
        mResMap["en"] = localeEn
        mResMap["ru"] = localeRu

        platformResources = MockResources(null)
        platformConfig = MockResourcesConfig("mockPlatform", platformResources)

        sourceConfig = MockSourceConfig("mockSource", mResMap)
        source = sourceConfig.resources

        config = Config(
            workDir = workDir,
            file = File(workDir, "config.json"),
            platform = platformConfig,
            source = sourceConfig,
            locales = setOf("en", "ru")
        )

        // Write config file to make it not empty
        config.file?.let {
            PrintWriter(it).run {
                write("{}")
                flush()
                close()
            }
        }
    }

    // ====================================================
    // Test invalid config

    @Test(expected = IllegalArgumentException::class)
    fun testNullPlatform() {
        LocoLaser.localize(config.copy(
            platform = null
        ))
    }

    @Test(expected = IllegalArgumentException::class)
    fun testNullSources() {
        LocoLaser.localize(config.copy(
            source = null
        ))
    }

    // ====================================================
    // Test conflict strategy

    @Test
    fun testConflictStrategyRemovePlatform() {
        val locale = "en"
        val newKey = "newKey"
        // Check if original map doesn't contain new key
        assertEquals(false, mResMap[locale]?.containsKey(newKey))

        val resLocale = ResLocale()
        resLocale.put(buildItem(newKey, "value"))

        platformResources.mMap = ResMap().apply {
            put(locale, resLocale)
        }

        assertTrue(LocoLaser.localize(config.copy(
            conflictStrategy = Config.ConflictStrategy.REMOVE_PLATFORM
        )))
        assertEquals(platformResources.mMap, source.mockMap)
    }

    @Test
    fun testConflictStrategyKeepNewPlatform() {
        val locale = "en"
        val newKey = "newKey"
        // Check if original map doesn't contain new key
        assertFalse(mResMap[locale]!!.containsKey(newKey))

        val overriddenKey = "key1"
        val overriddenValue = "valueEn1Overridden"
        // Check if original map contain overridden key with different value
        assertTrue(mResMap[locale]?.containsKey(overriddenKey) ?: false)
        assertNotEquals(overriddenValue, mResMap[locale]?.get(overriddenKey)?.valueForQuantity(
            Quantity.OTHER))

        val resLocale = ResLocale()
        resLocale.put(buildItem(newKey, "value"))
        resLocale.put(buildItem(overriddenKey, overriddenValue))

        platformResources.mMap = ResMap().apply {
            put(locale, resLocale)
        }

        assertTrue(LocoLaser.localize(config.copy(
            conflictStrategy = Config.ConflictStrategy.KEEP_NEW_PLATFORM
        )))
        assertNotEquals(platformResources.mMap, source.mockMap)

        platformResources.mMap?.get(locale)?.remove(newKey)
        assertEquals(platformResources.mMap, source.mockMap)
    }

    @Test
    fun testConflictStrategyKeepPlatform() {
        val locale = "en"
        val newKey = "newKey"
        // Check if original map doesn't contain new value
        assertFalse(mResMap[locale]?.containsKey(newKey) ?: false)

        val overriddenKey = "key1"
        val overriddenValue = "valueEn1Overridden"
        // Check if original map contain overridden key with different value
        assertTrue(mResMap[locale]?.containsKey(overriddenKey) ?: false)
        assertNotEquals(overriddenValue, mResMap[locale]?.get(overriddenKey)?.valueForQuantity(
            Quantity.OTHER))

        val resLocale = ResLocale()
        resLocale.put(buildItem(newKey, "value"))
        resLocale.put(buildItem(overriddenKey, overriddenValue))

        platformResources.mMap = ResMap().apply {
            put(locale, resLocale)
        }

        assertTrue(LocoLaser.localize(config.copy(
            conflictStrategy = Config.ConflictStrategy.KEEP_PLATFORM
        )))
        assertNotEquals(platformResources.mMap, source.mockMap)

        platformResources.mMap?.get(locale)?.remove(newKey)
        assertNotEquals(platformResources.mMap, source.mockMap)

        platformResources.mMap?.get(locale)?.get(overriddenKey)?.let {
            // Check if overridden value is not changed
            assertEquals(overriddenValue, it.valueForQuantity(Quantity.OTHER)?.value)

            // Check that it's only one difference
            source.mockMap?.get(locale)?.put(it)
            assertEquals(platformResources.mMap, source.mockMap)
        }
    }

    @Test
    fun testConflictStrategyExportNewLocal() {
        val locale = "en"
        val newKey = "newKey"
        // Check if original map doesn't contain new value
        assertFalse(mResMap[locale]!!.containsKey(newKey))

        val overriddenKey = "key1"
        val overriddenValue = "valueEn1Overridden"
        // Check if original map contain overridden key with different value
        assertTrue(mResMap[locale]?.containsKey(overriddenKey) ?: false)
        assertNotEquals(overriddenValue, mResMap[locale]?.get(overriddenKey)?.valueForQuantity(
            Quantity.OTHER))

        val resLocale = ResLocale()
        resLocale.put(buildItem(newKey, "value"))
        resLocale.put(buildItem(overriddenKey, overriddenValue))

        platformResources.mMap = ResMap().apply {
            put(locale, resLocale)
        }

        assertTrue(LocoLaser.localize(config.copy(
            conflictStrategy = Config.ConflictStrategy.EXPORT_NEW_PLATFORM
        )))
        assertEquals(platformResources.mMap, source.mockMap)

        source.mockMap?.get(locale)?.run {
            assertTrue(containsKey(newKey))
            assertTrue(containsKey(overriddenKey))
        }

        platformResources.mMap?.get(locale)?.get(overriddenKey)?.let {
            // Check if overridden value is replaced
            assertNotEquals(overriddenValue, it.valueForQuantity(Quantity.OTHER)?.value)
        }
    }

    @Test
    fun testConflictStrategyExportLocal() {
        val locale = "en"
        val newKey = "newKey"
        // Check if original map doesn't contain new value
        assertFalse(mResMap[locale]!!.containsKey(newKey))

        val overriddenKey = "key1"
        val overriddenValue = "valueEn1Overridden"
        // Check if original map contain overridden key with different value
        assertTrue(mResMap[locale]?.containsKey(overriddenKey) ?: false)
        assertNotEquals(overriddenValue, mResMap[locale]?.get(overriddenKey)?.valueForQuantity(
            Quantity.OTHER))

        val resLocale = ResLocale()
        resLocale.put(buildItem(newKey, "value"))
        resLocale.put(buildItem(overriddenKey, overriddenValue))

        platformResources.mMap = ResMap().apply {
            put(locale, resLocale)
        }

        assertTrue(LocoLaser.localize(config.copy(
            conflictStrategy = Config.ConflictStrategy.EXPORT_PLATFORM
        )))
        assertEquals(platformResources.mMap, source.mockMap)

        source.mockMap?.get(locale)?.run {
            assertTrue(containsKey(newKey))
            assertTrue(containsKey(overriddenKey))
        }

        platformResources.mMap?.get(locale)?.get(overriddenKey)?.let {
            // Check if overridden value is not changed
            assertEquals(overriddenValue, it.valueForQuantity(Quantity.OTHER)?.value)
        }
    }


    @Test
    fun testConflictStrategyExportMissedValues() {
        val missedKey = "key1"
        val locale = "ru"

        platformResources.mMap = ResMap(mResMap)
        assertNotNull(source.mockMap?.get(locale)?.remove(missedKey))

        assertTrue(LocoLaser.localize(config.copy(
            conflictStrategy = Config.ConflictStrategy.EXPORT_NEW_PLATFORM
        )))
        assertEquals(platformResources.mMap, source.mockMap)

        val missedItem = source.mockMap?.get(locale)?.get(missedKey)
        assertNotNull(missedItem)
    }

    // ====================================================
    // Other

    @Test
    fun testSimpleReadAndWrite() {
        assertNull(platformResources.mMap)
        assertNotNull(source.mockMap)

        assertTrue(LocoLaser.localize(config))

        assertEquals(platformResources.mMap, source.mockMap)
    }

    @Test
    fun testWritingConfigUsage() {
        assertNull(platformResources.mExtraParams)
        assertNotNull(config.extraParams)

        assertTrue(LocoLaser.localize(config))

        assertTrue(platformResources.mExtraParams == config.extraParams)
    }

    @Test
    fun testNoTrimUnsupportedQuantities() {
        val pluralItem = ResItem("value3").apply {
            addValue(ResValue("value ZERO", null, quantity = Quantity.ZERO))
            addValue(ResValue("value ONE", null, quantity = Quantity.ONE))
            addValue(ResValue("value TWO", null, quantity = Quantity.TWO))
            addValue(ResValue("value FEW", null, quantity = Quantity.FEW))
            addValue(ResValue("value MANY", null, quantity = Quantity.MANY))
            addValue(ResValue("value OTHER", null, quantity = Quantity.OTHER))
        }
        source.mockMap?.get("ru")?.put(pluralItem)
        source.mockMap?.get("en")?.put(pluralItem)

        assertTrue(LocoLaser.localize(config.copy(
            extraParams = ExtraParams().apply {
                putAll(config.extraParams)
                trimUnsupportedQuantities = false
            }
        )))

        assertEquals(
            ResItem("value3").apply {
                addValue(ResValue("value ZERO", null, quantity = Quantity.ZERO))
                addValue(ResValue("value ONE", null, quantity = Quantity.ONE))
                addValue(ResValue("value TWO", null, quantity = Quantity.TWO))
                addValue(ResValue("value FEW", null, quantity = Quantity.FEW))
                addValue(ResValue("value MANY", null, quantity = Quantity.MANY))
                addValue(ResValue("value OTHER", null, quantity = Quantity.OTHER))
            },
            platformResources.mMap?.get("ru")?.get("value3")
        )

        assertEquals(
            ResItem("value3").apply {
                addValue(ResValue("value ZERO", null, quantity = Quantity.ZERO))
                addValue(ResValue("value ONE", null, quantity = Quantity.ONE))
                addValue(ResValue("value TWO", null, quantity = Quantity.TWO))
                addValue(ResValue("value FEW", null, quantity = Quantity.FEW))
                addValue(ResValue("value MANY", null, quantity = Quantity.MANY))
                addValue(ResValue("value OTHER", null, quantity = Quantity.OTHER))
            },
            platformResources.mMap?.get("en")?.get("value3")
        )
    }

    @Test
    fun testTrimUnsupportedQuantities() {
        val pluralItem = ResItem("value3").apply {
            addValue(ResValue("value ZERO", null, quantity = Quantity.ZERO))
            addValue(ResValue("value ONE", null, quantity = Quantity.ONE))
            addValue(ResValue("value TWO", null, quantity = Quantity.TWO))
            addValue(ResValue("value FEW", null, quantity = Quantity.FEW))
            addValue(ResValue("value MANY", null, quantity = Quantity.MANY))
            addValue(ResValue("value OTHER", null, quantity = Quantity.OTHER))
        }
        source.mockMap?.get("ru")?.put(pluralItem)
        source.mockMap?.get("en")?.put(pluralItem)

        assertTrue(LocoLaser.localize(config.copy(
            extraParams = ExtraParams().apply {
                putAll(config.extraParams)
                trimUnsupportedQuantities = true
            }
        )))

        assertEquals(
            ResItem("value3").apply {
                addValue(ResValue("value ONE", null, quantity = Quantity.ONE))
                addValue(ResValue("value FEW", null, quantity = Quantity.FEW))
                addValue(ResValue("value MANY", null, quantity = Quantity.MANY))
                addValue(ResValue("value OTHER", null, quantity = Quantity.OTHER))
            },
            platformResources.mMap?.get("ru")?.get("value3")
        )

        assertEquals(
            ResItem("value3").apply {
                addValue(ResValue("value ONE", null, quantity = Quantity.ONE))
                addValue(ResValue("value OTHER", null, quantity = Quantity.OTHER))
            },
            platformResources.mMap?.get("en")?.get("value3")
        )
    }

    // ====================================================
    // Private classes and methods

    private fun buildItem(key: String, value: String): ResItem {
        val item = ResItem(key)
        item.addValue(ResValue(value, null))
        return item
    }

    private inner class MockResourcesConfig(
            override val type: String,
            override val resources: Resources
    ) : ResourcesConfig

    private inner class MockResources (
            var mMap: ResMap?
    ) : Resources {

        override val formattingType: FormattingType = NoFormattingType

        var mExtraParams: ExtraParams? = null

        override fun read(locales: Set<String>?, extraParams: ExtraParams?): ResMap {
            val resMap = ResMap()
            mMap?.let { map ->
                locales?.forEach { locale ->
                    resMap[locale] = ResLocale(map[locale])
                }
            }
            return resMap
        }

        @Throws(IOException::class)
        override fun write(resMap: ResMap, extraParams: ExtraParams?) {
            mMap = ResMap(resMap)
            mExtraParams = extraParams
        }

        override fun allFiles(locales: Set<String>): List<File> {
            throw RuntimeException("This method shouldn't be called")
        }
    }

    private class MockSourceConfig (
            override val type: String,
            resMap: ResMap
    ) : ResourcesConfig {
        override val resources: MockSource = MockSource(ResMap(resMap))
    }

    private class MockSource(
        var mockMap: ResMap?
    ) : Resources {

        override val formattingType: FormattingType = NoFormattingType

        override fun read(locales: Set<String>?, extraParams: ExtraParams?): ResMap? {
            return ResMap(mockMap)
        }

        override fun write(resMap: ResMap, extraParams: ExtraParams?) {
            if (mockMap == null)
                mockMap = ResMap(resMap)
            else
                mockMap.merge(resMap)
        }

        override fun allFiles(locales: Set<String>): List<File> {
            throw RuntimeException("This method shouldn't be called")
        }
    }
}
